module/zfs/zfs_ioctl.c

```diff
@@ -5232,25 +5499,992 @@ zfs_ioc_send_space(const char *snapname, nvlist_t *innvl, nvlist_t *outnvl)
+	.zvec_secpolicy		= zfs_secpolicy_inherit_prop,
+	.zvec_namecheck		= DATASET_NAME,
+	.zvec_pool_check	= POOL_CHECK_SUSPENDED | POOL_CHECK_READONLY,
+	.zvec_smush_outnvlist	= B_FALSE,
+	.zvec_allow_log		= B_TRUE,
+},
+{	.zvec_name		= "zfs_destroy",
+	.zvec_func		= zfs_stable_ioc_zfs_destroy,
+	.zvec_secpolicy		= zfs_secpolicy_destroy,
+	.zvec_namecheck		= DATASET_NAME,
+	.zvec_pool_check	= POOL_CHECK_SUSPENDED | POOL_CHECK_READONLY,
+	.zvec_smush_outnvlist	= B_FALSE,
+	.zvec_allow_log		= B_TRUE,
+},
+};
+
+static const ssize_t zfs_stable_ioc_vec_count =
+    sizeof (zfs_stable_ioc_vec) / sizeof (zfs_stable_ioc_vec_t);
+
+/* Helper function to validate names */
+static inline int
+zfs_namecheck(const char *name, zfs_ioc_namecheck_t namecheck,
+    zfs_ioc_poolcheck_t pool_check) {
+	int error = 0;
+
+	switch (namecheck) {
+	case POOL_NAME_OR_EMPTY:
+		if (name[0] == '\0')
+			break;
+	case POOL_NAME:
+		if (pool_namecheck(name, NULL, NULL) != 0)
+			error = SET_ERROR(EINVAL);
+		else
+			error = pool_status_check(name,
+			    POOL_NAME, pool_check);
+		break;
+
+	case DATASET_NAME_OR_EMPTY:
+		if (name[0] == '\0')
+			break;
+	case DATASET_NAME:
+		if (name[0] == '\0')
+			break;
+		if (dataset_namecheck(name, NULL, NULL) != 0)
+			error = SET_ERROR(EINVAL);
+		else
+			error = pool_status_check(name,
+			    DATASET_NAME, pool_check);
+		break;
+
+	case NO_NAME:
+		break;
+	}
+
+	return (error);
+}
+
+/*
+ * Stable API
+ * We use an XDR-encoded nvlist for input, which is encapsulated inside the
+ * legacy zfs_cmd_t.
+ *
+ * The format is:
+ *
+ * zc->zc_name -> dataset/pool name
+ * zc->zc_nvlist_src: {
+ *     "cmd" -> command name (string)
+ *     "version" -> command version (uint64)
+ *     "opts" -> command specific options (nvlist)
+ *     "innvl" -> command specific input data
+ * }
+ *
+ * Malformed XDR-encoded nvlists will result in get_nvlist() returning an error
+ * from nvlist_unpack(), which fails early. Consequently, all strings can be
+ * safely assumed to be NULL terminated.
+ *
+ * Names of commands focused on the DSL are prefixed with "zfs_" while names of
+ * commands focused on the SPA are prefixed with "zpool_".
+ *
+ * XXX: Unit tests are necessary, especially for malformed XDR-encoded nvlist
+ * inputs.
+ */
+static int
+zfs_ioc_stable(zfs_cmd_t *zc)
+{
+	const char *name = zc->zc_name;
+	char *cmd;
+	nvlist_t *mnvl, *innvl, *outnvl, *opts;
+	uint64_t version;
+	int error = 0;
+	ssize_t i;
+
+	zc->zc_name[sizeof (zc->zc_name) - 1] = '\0';
+
+	if (zc->zc_nvlist_src_size != 0) {
+		error = get_nvlist(zc->zc_nvlist_src, zc->zc_nvlist_src_size,
+		    zc->zc_iflags, &mnvl);
+		if (error != 0)
+			goto out;
+	}
+
+	if ((error = nvlist_lookup_string(mnvl, "cmd", &cmd)))
+		return (error);
+
+	if ((error = nvlist_lookup_nvlist(mnvl, "innvl", &innvl)))
+		return (error);
+
+	if (nvlist_lookup_nvlist(mnvl, "opts", &opts))
+		opts = NULL;
+
+	if ((error = nvlist_lookup_uint64(mnvl, "version", &version)))
+		return (error);
+
```

Comment:
General cstyle nit:
When using a non-boolean expression in a condition it's preferred to explicitly compare to 0.
`if ((error = ...) != 0)`
`if (nvlist_lookup_nvlist(...) != 0)`
etc.